crates/wysiwyg/src/composer_model/hyperlinks.rs

```diff
@@ -58,9 +61,63 @@ where
 
     pub fn set_link(&mut self, link: S) -> ComposerUpdate<S> {
         self.push_state_to_history();
-        let (s, e) = self.safe_selection();
-        let range = self.state.dom.find_range(s, e);
-        self.set_link_range(range, link)
+        let (mut s, mut e) = self.safe_selection();
+
+        let mut range = self.state.dom.find_range(s, e);
+
+        // Find container link that completely covers the range
+        if let Some(link) = self.find_parent_links(&range) {
+            // If found, update the range to the container link bounds
+            range = self.state.dom.find_range_by_node(&link);
+            (s, e) = (range.start(), range.end());
+        }
+
+        if s == e {
+            return ComposerUpdate::keep();
+        }
+
+        let inserted = self
+            .state
+            .dom
+            .insert_parent(&range, DomNode::new_link(link, vec![]));
+
+        // Ensure no duplication by deleting any links contained within the new link
+        self.delete_child_links(&inserted);
+
+        self.create_update_replace_all()
+    }
+
+    fn delete_child_links(&mut self, node_handle: &DomHandle) {
+        let node = self.state.dom.lookup_node(node_handle);
+
+        for handle in node
+            .iter_containers()
+            .filter(|n| matches!(n.kind(), ContainerNodeKind::Link(_)))
+            .map(|n| n.handle())
+            .filter(|h| *h != *node_handle)
+            .collect::<Vec<_>>()
+            .into_iter()
+            .rev()
+        {
+            self.state.dom.remove_and_keep_children(&handle);
+        }
+    }
+
+    fn find_parent_links(&mut self, range: &Range) -> Option<DomHandle> {
```

Comment:
This could be `find_closest_ancestor_link`? Since it returns an `Option<DomHandle>`, the name seems a bit misleading.

Reply:
[`8180d8d`](https://github.com/matrix-org/matrix-rich-text-editor/pull/423/commits/8180d8d1a3a92863137bdbead0df6c445c768b96)